<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en" lang="en-us">
<head>
   <meta http-equiv="content-type" content="text/html; charset=utf-8" />
   <title>Git Patch简单教程</title>
   <meta name="author" content="Peer Xu" />
   <link href="http://peerxu.github.com/" rel="alternate" title="Peer Studio Blog" type="application/atom+xml" />

   <!-- syntax highlighting CSS -->
   <link rel="stylesheet" href="/css/syntax.css" type="text/css" />

   <!-- Homepage CSS -->
   <link rel="stylesheet" href="/css/screen.css" type="text/css" media="screen, projection" />

	<!-- Google Analytics -->
	<script type="text/javascript">

	  var _gaq = _gaq || [];
	  _gaq.push(['_setAccount', 'UA-26349380-3']);
	  _gaq.push(['_trackPageview']);

	  (function() {
		var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
		ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
		var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
	  })();

	</script>
	<!-- Google Analytics end -->
</head>
<body>

<div class="site">
  <div class="title">
    <a href="/">Peer Studio Blog</a>
    <a class="extra" href="/">home</a>
  </div>
  
  <div id="post">
<h1>Git Patch简单教程</h1>

<p>今天我们来学习一下关于git怎么创建与应用patch.</p>

<h3>预备资料</h3>

<p><a href="http://zh.wikipedia.org/zh-cn/Patch">patch</a>是用来对现有的源代码进行更新的一种手段.</p>

<p>patch可以通过diff, git, svn, hg等工具生成.</p>

<p>diff与git产成的patch文件有点差异(下面会简单的介绍patch文档的结构).</p>

<p>下面我们就来介绍怎么创建一个patch.</p>

<h3>前戏</h3>

<p>首先, 你得有一个好环(ji)境(you), 才能开展patch的工作.</p>

<p>所以我在这里简单的创(ji)建(you)一个环境给大家使用. :P</p>

<pre><code>$ mkdir Hello
$ cd hello
$ git init
$ cat &lt;&lt; EOF &gt; hello.py
#!/usr/bin/env python2.7
print "hello, world"
EOF
$ git add .
$ git commit -m "initial commit" -a
</code></pre>

<p>至此, 大家已经有一个可以使用的好环(ji)境(you).(怎么感觉节奏不对)</p>

<h3>动作片播放ing...</h3>

<p>好了, 我们已经创建好了一个可以用的基友, 那么就开始干吧!</p>

<pre><code>$ git checkout -b my-feature
$ vi hello.py
$ cat hello.py
#!/usr/bin/env python2.7
def hello(name):
    print "hello, %s" % name

hello("world")
$ git commit -m "add feature" -a
$ git format-patch master
$ ls
0001-add-feature.patch  hello.py
</code></pre>

<p>下面来为大家讲解一下上面的<em>动作</em>要领.</p>

<p>首先我们很大众的开局, 先创建一个my-feature.</p>

<p>然后啪啪啪地编辑了一下文档, 添加了一个相当犀利的特性.</p>

<p>然后按照正常套路去commit咱们牛逼的特性.</p>

<p>当全世界都以为我们要merge到master的时候, 我只能说你们, 图样图森普.</p>

<p>我们的目标是没有蛀牙. 咳咳, 是创建patch.</p>

<p>所以我们最后使用 <code>git format-patch master</code> 创建一个patch.</p>

<p>好了, 通常动作片的中场都是要休息一下的.</p>

<h3>动作片继续播放ing...</h3>

<p>我们既然已经生成了patch文件, 咱们就看看这文件到底是什么样子的.</p>

<pre><code>From a9706e72ba4976f69d1c3ca5e93ab9a00757c9fd Mon Sep 17 00:00:00 2001
From: Peer Xu &lt;pppeerxu@gmail.com&gt;
Date: Thu, 21 Nov 2013 23:37:41 +0800
Subject: [PATCH] add feature

---
 hello.py | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/hello.py b/hello.py
index 6635083..c314e13 100644
--- a/hello.py
+++ b/hello.py
@@ -1,2 +1,5 @@
 #!/usr/bin/env python2.7
-print "hello,world"
+def hello(name):
+    print "hello, %s" % name
+
+hello("world")
--
1.8.3.2
</code></pre>

<p>以我阅片无数的经验, 这是一封邮件.</p>

<p>嗯, 剩下的就留给各位看官自己去YY吧.</p>

<h3>动作片通常都是分开三部播放的...</h3>

<p>这里, 我传授大家怎么将git生成的patch应用到项目中去.</p>

<p>还是咱们的好环境.</p>

<pre><code>$ cp 0001-add-feature.patch /dev/shm/
$ rm 0001-add-feature.patch # 删除生成的patch, 保证环境干净 (这是你洁癖的证据!)
$ git checkout master
$ git am /dev/shm/0001-add-feature.patch
$ cat hello.py
#!/usr/bin/env python2.7
def hello(name):
    print "hello, %s" % name

hello("world")
$ git log # 这里可以看到应用patch的日志
</code></pre>

<p>我们先将patch复制到另外一个地方(假装是从很远的地方下载下来的)</p>

<p>然后checkout到master(假装在别的环境中...)</p>

<p>然后使用 <code>git am</code> 这命令使patch应用到环境中去.</p>

<p>然后, 就没有然后了...</p>

<h3>动作片再好, 也不能经常看...</h3>

</div>

<div id="related">
  <h2>Related Posts</h2>
  <ul class="posts">
    
      <li><span>07 Nov 2014</span> &raquo; <a href="/2014/11/07/dns-hijack-with-dnsmasq.html">一次无意识的DNS劫持</a></li>
    
      <li><span>21 Aug 2014</span> &raquo; <a href="/2014/08/21/sqlalchemy-many-to-many-relationship.html">SQLAlchemy多对多关系使用方法</a></li>
    
      <li><span>08 Jul 2014</span> &raquo; <a href="/2014/07/08/choose-your-favorite-number-from-alibaba.html">老板, 要挑个号码么?</a></li>
    
  </ul>
</div>

  
  <div class="footer">
    <div class="contact">
      <p>
        Peer Studio<br />
		Weibo<br />
		Twitter<br />
        Cofounder of <a href="http://github.com/">GitHub</a><br />
		pppeerxu@gmail.com
      </p>
    </div>
    <div class="contact">
      <p>
        <a href="http://github.com/peerxu/">github.com/PeerXu</a><br />
		<a href="http://weibo.com/peerstudio">weibo.com/peerstudio</a><br />
		<a href="http://twitter.com/peerxu">twitter.com/PeerXu</a><br />
      </p>
    </div>
    <div class="rss">
      <a href="atom.xml">
        <img src="/images/rss.png" alt="Subscribe to RSS Feed" />
      </a>
    </div>
  </div>
</div>

<a href="http://github.com/peerxu"><img style="position: absolute; top: 0; right: 0; border: 0;" src="http://s3.amazonaws.com/github/ribbons/forkme_right_red_aa0000.png" alt="Fork me on GitHub" /></a>

</body>
</html>
